File: 프로그래머스/2/178870. 연속된 부분 수열의 합/연속된 부분 수열의 합.js
function solution(sequence, k) {
    let start = 0, end = 0;
    let currentSum = sequence[0];
    let minLen = sequence.length;
    let answer = [0, sequence.length - 1];

    while (start < sequence.length && end < sequence.length) {
        if (currentSum < k) {
            end++;
            if (end < sequence.length) currentSum += sequence[end];
        } else if (currentSum > k) {
            currentSum -= sequence[start];
            start++;
        } else { // currentSum === k
            if ((end - start) < minLen) {
                minLen = end - start;
                answer = [start, end];
            }
            currentSum -= sequence[start];
            start++;
        }
    }

    return answer;
}

File: 프로그래머스/2/178870. 연속된 부분 수열의 합/연속된 부분 수열의 합.kt
class Solution {
    fun solution(sequence: IntArray, k: Int): IntArray {
        val n = sequence.size
        var start = 0
        var end = 0
        var currentSum = sequence[0]

        var minLen = n
        var answer = intArrayOf(0, n - 1) // 초기값: 최대 길이 구간

        while (start < n && end < n) {
            when {
                currentSum < k -> {
                    end++
                    if (end < n) currentSum += sequence[end]
                }
                currentSum > k -> {
                    currentSum -= sequence[start]
                    start++
                }
                else -> { // currentSum == k
                    if ((end - start) < minLen) {
                        minLen = end - start
                        answer = intArrayOf(start, end)
                    }
                    currentSum -= sequence[start]
                    start++
                }
            }
        }

        return answer
    }
}